runtime/runtime.go

```diff
@@ -50,7 +57,9 @@ type provider struct {
 	coe
 	ExtractedCredential
 	postSandboxCreate  []PostSandboxCreate
+	preSandboxCreate   []PreSandboxCreate
```

Comment:
We have multiple functions provided as types. I'd rather use an interface ```SandboxHooks``` and encapsulate all of them.

Reply:
Exactly like ```ExtractedCredential```.
```go
type SandBoxHooks interface {
     PostSandboxCreate(...)
     PreSandboxCreate(...)
     PostSandboxDestroy(...)
     PreSandboxDestroy(...)
}
```

Reply:
But that is not how this works right, we want to add an arbitrary number of hooks? I don't understand how this interface works in the context of how we are using the hooks.

Reply:
```go
type SandBoxHooks interface {
     postSandboxCreate(...)
     preSandboxCreate(...)
     postSandboxDestroy(...)
     preSandboxDestroy(...)
}

type Configuration struct {
        SandBoxHooks
	ExtractedCredential
}

Func NewRuntime(config Configuration) {

        ...

        var c ExtractedCredential
        if config.ExtractedCredential == nil 
                c = defaultExtractedCredential{}
        } else {
                c = config.ExtractedCredential
        }

        var h SandBoxHooks
        if config.SandBoxHooks == nil 
                h = cluster.SandBoxHooks
        } else {
                h = config.SandBoxHooks
        }

        Provider = &provider{coe: cluster, ExtractedCredential: c, Hooks: h}
}

func (p provider) CreateSandbox(...) {

...
         p.Hooks.postSandboxCreate(...)
...
}
```

Reply:
The crux of this argument comes down to how users primarily use this API.  Will they go with what a provider has then append to it or do they want to overwrite everything. Right now, we only have one hook and it's in the openshift provider so appending is fine.  To me, this solution is functional, but I think we could come back to this if we hit a limitation with append, but it's too soon to tell if we will.  

My favorite option is still to use the interface and maybe create a core set of hooks enabled by flags, but I'll go with what folks think on this.  I don't want to block this work from proceeding.